### ISSUE TITLE
Implement PCM voices support through YM2612's DAC

### ISSUE THREAD

**Issue by Wohlstand** [Wohlstand/libOPNMIDI#5]
As YM2612 supports DAC channel that can be used to output PCM, it was widely used for realistic percussion, SFX, and other things. DAC also can be used. No need to dedicate multiple DAC channels on multiple chip emulators, needed only one on any chip to don't steal so expressive FM channels from every simulated chip.

**Comment by papiezak** [comment]
I have been thinking about it, just one question - do you plan staying close to limitations of Mega Drive hardware (Z80 CPU speed, RAM size, DMA, etc.) ?
If not, then, well... YM2612's DAC enable feature is entirely software controlled, so via software mixing you may have even like 8-bit 64 PCM channels @44.1KhZ with advanced volume and pitch control + maybe some after effects (all of this just from one hardware DAC channel) as long as you have enough CPU power.

**Comment by Wohlstand** [comment]
For the DAC is planned to provide some sort of simple WaveTable synth where it's output will be played through DAC. I have plan to keep 8-bit PCM output, but the WT-Synth backend possibly will don't rely on Z80, but will give only some basic percussion support with minimalistic functionality. I have to give it two modes: 
* **FAKE-DAC**, which means the output of WT-Synth will NOT be sent into emulator's DAC interface, but will be directly mixed with output of emulated chips. This will save lot of CPU power. This way is only possible to use also with OPL3 chip which doesn't have any DAC. So, with this will be possible to simulate JunleVision's Sound Drive demo where OPL3 FM is mixed with PCM drums played through audio output interface of the sound card.
* **Real-DAC**, which means the output of WT-Synth will be directly sent to emulator's DAC interface. As this going to be sample-by-sample, then it's potential increasing of CPU usage due increasing of the data transfer complexity. This way would allow to even use some sort of Arduino rig to output the generated stuff into a real chip (or multi-chip in dependence from implementation you will construct).

**Comment by papiezak** [comment]
I see, but 
>will give only some basic percussion support with minimalistic functionality.

Will the be possibility of handling SFX instruments, orchestra hits or timpani in melodic banks? I know OPN2 DAC requires a lot of software mixing in order to be used in a musical way as pitch control isn't natively supported, but then again...

**Comment by Wohlstand** [comment]
Some sort of melodics are will be too, but at first there are will have no envelope support. Everything will be done software-side. The DAC itself is a silly receiver of PCM samples sequence. To play the sound through DAC with specific sample rate, you will must to keep same delay between of each sample until write it into the DAC port.

**Comment by Wohlstand** [comment]
Just for a simple test, I have made the super-simple raw PCM player through Nuked OPN2's DAC: https://github.com/Wohlstand/libOPNMIDI/commit/77d0d5092ba87573601ee52110bc59edb765c353, and it works! :smile:

**Comment by Wohlstand** [comment]
Damn, I forgot to declare "SDL_MAIN_HANDLED", it's SDL's header has declared it's SDL_main...

**Comment by Wohlstand** [comment]
but... Yeah! it's really need to use it...

**Comment by Wohlstand** [comment]
@papiezak , Just now I have sent a fix!

**Comment by papiezak** [comment]
Yeah, it compiles, but it crashes the instant I load your song

**Comment by Wohlstand** [comment]
Just now sent another fix, please retry...

**Comment by papiezak** [comment]
Sorry, still crashing

**Comment by Wohlstand** [comment]
Try to compile it in debug build (`-DCMAKE_BUILD_TYPE=Debug`), and run it through GDB:
```
gdb --args dac_test "Path\To\RAW\file"
```
then:
```
(gdb) run
```
and when crash will happen, type:
```
(gdb) bt
```
and give me the screenshot

**Comment by Wohlstand** [comment]
I think I know why this happens... in the code I forgot to initialize Nuked's instance after allocation :man_facepalming:

**Comment by Wohlstand** [comment]
@papiezak , just now I have pushed my fix, try it out please :wink:

**Comment by papiezak** [comment]
I will try tomorrow

**Comment by Wohlstand** [comment]
Okay, anyway, I hope, it's now fixed, or I'll try that on Windows by myself now...

**Comment by papiezak** [comment]
yeah it's alright! playback doesn't jitter, great work :1st_place_medal: 

don't worry about 44.1 khz, real ym2612 can do it as well, you just need either Fujitsu FM Towns (16 MHz 386 of most basic models should do) or overclocked mega drive: http://www.sega-16.com/forum/showthread.php?20795-YM2612-playing-8-Bit-PCM-at-44-1-khz!&s=6159e5ef7714d5d1889a5b44e6c5e333

**Comment by Wohlstand** [comment]
I think, the sample rate of played stream must be chosen optimal, especially if 44100 is looks hacky for the real chip. And I'll need to make this test also play something from FM too. As in this test I have played PCM only and all other FM channels are muted.

**Comment by Wohlstand** [comment]
@jpcima , you can join the discussion here. Anyway, the first experiment with WT-Synth I want to make it on OPN2 Bank Editor side to provide the ability to upload WAV files, convert them into U8 and necessary sample rate, and then give the UI to configure the sample and test it's playing. I'll try to create the super-simple WT-Synth which we can extend with resampler (to scale the pitch), and you can try out your existing things you have suggested me when we talked via XMPP.

**Comment by papiezak** [comment]
>I think, the sample rate of played stream must be chosen optimal, especially if 44100 is looks hacky for the real chip.

26 KHz is optimal for real hardware, Deflemask and VGM music maker has 32 Khz option though. Most games used from 8 to circa 19 Khz

**Comment by jpcima** [comment]
What makes the Fake-DAC fundamentally different from playing OPNMIDI with a Fluidsynth instance layered on top?

**Comment by Wohlstand** [comment]
The difference is that will be used same simplified WT-Synth and the simulation of DAC sound processing (the u8 -> s16 conversion formula can be directly stolen from Nuked OPN2 emulator's code to prepare the buffer to mix with the generated FM outputs). With the same success result we are able to stream FluidSynth into DAC directly with requested sample rate and u8 mono format.

**Comment by jpcima** [comment]
What is the added value of this? Under real-dac, I understood a signal can be mixed as carrier into the FM algorithm arrangement.

But fake-dac? It's nothing special to add two ouputs of synths together. Beside, why is the insistence to tie it to the wavetable concept? It could imagine it to be any kind of generator, including from audio port input. 

I use as example an analog Moog phatty synth, where there is a kinda similar concept: an input jack port is for an externally generated oscillator 3, which goes to be processed in the mixing and filter stages. I can plug a this jack to any synth of my liking.
(plugging the Moog synth into YM2612 would be a fun experiment to try definitely!)

**Comment by Wohlstand** [comment]
The purpose of Fake-DAC is only reducing the complexity of the PCM playback with emulators (in comparison with sample-by-sample posting into DAC interface of the emulated chip) and giving ability to optionally use it with OPL3 chip for some tweaks (as OPL3 has no DAC, it's the only option to stream PCM independently but synchronously).

> But fake-dac? It's nothing special to add two ouputs of synths together. Beside, why is the insistence to tie it to the wavetable concept? It could imagine it to be any kind of generator, including from audio port input.

Through ANY way (Fake-DAC that is playing separately and mixing with post-generated output of all chips, or Real-DAC which is streaming directly into DAC interface on one of simulating chips) is possible to play anything with no matter how that will be synthesized or decoded. 

> I use as example an analog Moog phatty synth, where there is a kinda similar concept: an input jack port is for an externally generated oscillator 3, which goes to be processed in the mixing and filter stages. I can plug a this jack to any synth of my liking.
(plugging the Moog synth into YM2612 would be a fun experiment to try definitely!)

Just for some experiments it's would be nice to try and listen the result :smile:

**Comment by jpcima** [comment]
This is my personal idea about realizing this process.

- The "Fake-DAC": allow a DAC-enabled channel to transfer its events to MIDI hook functions. So, a synthesizer instance can receive and synthesize a channel in parallel (which can be Fluidsynth, simple WT, hardware, etc). Mixing can be fully external so libOPNMIDI doesn't have to care.

- Real-DAC: MIDI hook like above + frame generator hook set up by caller. The hook asks for the next frame to feed in the emulator, and there is not a significant increase of complexity. It leaves all the responsability of managing synth complexity on the user side.

**Comment by Wohlstand** [comment]
Just now I have updated DAC demo and I have allowed it to play files of different sample rates (every sample is passed to the chip with the delay declared by sample rate ratio between of native and input stream).

[raw_pcm_u8_different_rates.zip](https://github.com/Wohlstand/libOPNMIDI/files/2091842/raw_pcm_u8_different_rates.zip)

**Comment by jpcima** [comment]
About the real DAC idea, some questions.

- when the midi channel has a DAC-enabled instrument, it has channels 1-5 as FM and 6 as special.
  as the normal notes are dispatched to the channels 1-5, do at the same time _all_ of them are sent to the DAC master of channel 6? (being a synth of some kind)

- how is decided the ownership of the unique DAC channel, when banks have several instruments which demand the use of DAC?
  at what point is decided a channel has stopped using DAC, and ownership transfers to another?

**Comment by Wohlstand** [comment]
* As here is multichip case, yeah, enabling one DAC on a first chip and stream into it. Then, the channel categories thing can limit FM writing into that channel
* We are will use one DAC channel of one chip to stream mixed output. It's not need to have something extra: when we are using DAC-enabled instruments, we are turning DAC channel, and keeping it be always on.

**Comment by Wohlstand** [comment]
As option, I want to make the instrument have both FM and DAC implementations to be able have the "FM-Only" option that will disable usage of DAC and tell using FM implementation of instrument [which will be kept as fallback for this case].

**Comment by jpcima** [comment]
I have insisted on such a format as sfz for such reason: I want you take a look and evaluate.
You don't need an external library to make this format work, nor editor.

A sfz is next to trivial for saving and loading process. A text editor can write it.
You can easily make an editor into OPN2-BE which saves into this format, even it just 1% of specification.
An example will show you how simple is a hello world sample: https://en.wikipedia.org/wiki/SFZ_(file_format)

[Here](https://www.linuxsampler.org/sfz/) will show you all ways it can be specified, and how all perspective of extending WT will be already existing. Imo, WOPN cannot possibly offer such an expressivity as a good sampler can have, in any short term work.

In such way, and as was my point I defended, one can implement a minimum WT synth, and have all room for adding support a few at a time.

The only thing, this format is not self-contained, but as I'm concerned, I don't find it an inconvenience to carry multiple files, rather that one large multi-megabyte file wopn.

**Comment by Wohlstand** [comment]
Checked out the WOPNv3 draft and yeah, about of PCM side:
- instrument supports only one PCM chunk (it's would be nice to support many of them with range allocation like done in most of WT synths. That needed for melodic voices, not for drums)
- instrument supports only one layer (can't have multiple PCM chunks are playing in parallel)

Have to improve the format with adding of layers and ranges per chunk

**Comment by Wohlstand** [comment]
About of sfz: I'll adapt the internal store for sfz thing. But yeah, as WOPNv3 gives the support of difference instrument data in dependence on the header, then yeah, it's possible to put absolutely anything on the instrument part.

Then yeah, it's would be next key=value store in next format:
```
uint16be | count of entries
{
  1         | type
  NT-String | key
  <sz>      | raw data in dependence on a given type
}
```

**Comment by Wohlstand** [comment]
The V2 format is still fine as mainstream even lacks those flags are unused yet, the V3 can be added as "experimental" while the development process will go. Anyway, as I said, the format allows to don't include parts are not specified. I.e. the same V3 is able to be saved with no any PCM-related stuff while it is WIP. I did the variadic size support per instrument where data blocks are not writing into file when there are disabled to reduce size of the file (I am even able to save blank instruments as 1-byte entries!!!) and also, the WIP PCM part can be changed on the side after main release.

**Comment by igdegoo** [comment]
this still being looked into?

**Comment by Wohlstand** [comment]
Yes, it's is! For now the work on the side of OPN2-BE to implement support into bank files to make PCM-based instruments be stored into them.

**Comment by freq-mod** [comment]
As PCM sample playback issue keeps appearing, I would like to propose something:
>No need to dedicate multiple DAC channels on multiple chip emulators, needed only one on any chip to don't steal so expensive FM channels from every simulated chip.

Idk if it's good idea. Most will use PCM DAC for drumkits and one channel won't be enough, especially with GS/XG midis. So:
- Ch6 PCM mode can be enabled for any chip, most devices can handle at last 6 or so MAME 2612 cores, so FM channels count isn't bothersome.
- There can be a fallback option - an instrument by default would play PCM sample, but in case there are no PCM channels available, it would fallback to FM synthesis alternative.

also YM2608 rhythm rompler can be supported for a start, it can't be hard

**Comment by freq-mod** [comment]
Very important @Wohlstand - max sample rate of 2612 PCM is tied to chip clock rate, at md's rate of 7.67 mhz it is 26633 hz (chip sample rate/2), st higher rates it loses writes to the register

**Comment by Wohlstand** [comment]
As I noticed at the comment to the commit, DAC has a serious limitation on the maximum volume (it's 6x smaller than it could be, just it was made being same volume as every FM voice). And this sets a serious limit on how loud PCM-played drums will sound, or limit single instrument at the DAC, or make voices even more quiter to fit into the limitation.

Some VGM samples I made with these experiments:

- [human3-gm.zip](https://github.com/user-attachments/files/26335285/human3-gm.zip)
- [Dancing Willow - Drowsy Maggie.zip](https://github.com/user-attachments/files/26335286/Dancing.Willow.-.Drowsy.Maggie.zip)
- [CAVE3-loop-2ports-1.MID.zip](https://github.com/user-attachments/files/26335287/CAVE3-loop-2ports-1.MID.zip)
- [briefing-gm.zip](https://github.com/user-attachments/files/26335288/briefing-gm.zip)
- [393.zip](https://github.com/user-attachments/files/26335289/393.zip)
- [384.zip](https://github.com/user-attachments/files/26335290/384.zip)
- [383 - o-game.zip](https://github.com/user-attachments/files/26335291/383.-.o-game.zip)
- [382.zip](https://github.com/user-attachments/files/26335292/382.zip)
- [380.zip](https://github.com/user-attachments/files/26335295/380.zip)
- [Wally.zip](https://github.com/user-attachments/files/26335296/211.zip)

And you might hear clipping of the DAC when two voices gets played together.

**Comment by freq-mod** [comment]
then soft-mixing code is imperfect, quite a handful of games used up to 4ch PCM with minimal to no clipping. 

https://www.youtube.com/watch?v=fZWwv5FNLE0 if it heps

**Comment by Wohlstand** [comment]
I'll check this later, I am unable to watch YouTube outside my home network (I am at the office). Do you have a textual _article_ that explains the algorithm?

**Comment by freq-mod** [comment]
Sadly, vid doesnt go too in depth for mixing channels - it boils down to half each sample values and add them up. Also, a volume scaling based a LUT pitch table is applied.

There is a DualPCM technical paper (a manual for community-made 2ch PCM driver for Mega Drive, which 1. doesn't clip, 2. doesn't jitter, 3. can repitch samples) - https://mrjester.hapisan.com/05_DualPCM/Dual%20PCM%20-%20FlexEd%20(Technical%20Document).pdf Excerpt:

> Volume control
This part of the driver is more complicated, so understand the below is a description of the volume control and how it
works, for details of using volume control, please see the next chapter. I do recommend at least glossing over this page
for your information.
Many methods were tried and practiced with Dual PCM to find the right volume control method, after experiments and
discussions with my testers, we have come to the conclusion that there is no single best way to go about this. We have
three main methods all of which are similar in practice, some have benefits in one way but downfalls in another, of which
other methods will have the reverse of.
The main idea is the same throughout all three methods. Dual PCM has a volume look up table (VLUT), one for each
virtual channel. This is a block of $100 bytes, these bytes will swap in substitution with the actual samples bytes to
simulate a volume difference. Effectively speaking, the samples are being re-sampled with a different volume in real-time
right before they are played.
These two $100 byte tables are actually "buffers", so you can swap out the volume table for a different table which
contains completely different volume values. But it's the swapping out that's different between the methods.

**Comment by Wohlstand** [comment]
Looks interesting, I'll try to make some once I get my home.

**Comment by Wohlstand** [comment]
Okay, I red the PDF, and it looks interesting, however, for now I am going to sleep. I also spent a time to fix a serious bug at the libGME that was been reported to me via email.
